### PR TITLE
New module: `plink2/extract`

### LINKED ISSUE #1227: new module: plink2/extract
I think it would be good to have a module for [`plink2/extract`](https://bioconda.github.io/recipes/plink2/README.html). plink2 introduces a new file format (pgen / pvar / psam) which replaces the traditional bfile format in plink1 (bed / bim / fam).

- [X] This module does not exist yet with the [`nf-core modules list`](https://github.com/nf-core/tools#list-modules) command
- [X] There is no [open pull request](https://github.com/nf-core/modules/pulls) for this module
- [X] There is no [open issue](https://github.com/nf-core/modules/issues) for this module
- [X] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by nebfield**
## PR checklist

Closes #1227. plink2 introduces a new file format (pgen / pvar / psam) which replaces the traditional bfile format in plink1 (bed / bim / fam). This submitted module complements the existing `plink/extract` module. 

- [X] This comment contains a description of changes (with reason).
- [X] If you've fixed a bug or added code that should be tested, add tests!
- [X] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [X] If necessary, include test data in your PR.
- [X] Remove all TODO statements.
- [X] Emit the `versions.yml` file.
- [X] Follow the naming conventions.
- [X] Follow the parameters requirements.
- [X] Follow the input/output options guidelines.
- [X] Add a resource `label`
- [X] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [X] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [X] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [X] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

File: modules/plink2/extract/main.nf [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    output:
+    tuple val(meta), path("*.pgen"), emit: pgen
+    tuple val(meta), path("*.psam"), emit: psam
+    tuple val(meta), path("*.pvar"), emit: pvar
+    path "versions.yml"            , emit: versions
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    if( "$pgen" == "${prefix}.pgen" ) error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"
```
Comment (jfy133) [inline modules/plink2/extract/main.nf:22]:
```suggestion
    if( "$pgen" == "${prefix}.pgen" ) error "Input and output names are the same, use \"task.ext.prefix\" in modules.config to disambiguate!"
```

File: modules/plink2/extract/main.nf [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    """
+    plink2 \\
+        --pfile ${pgen.baseName} \\
+        $args \\
+        --threads $task.cpus \\
+        --extract $variants \\
+        --make-pgen \\
+        --out ${prefix}
+
+    echo $task.memory
```
Comment (jfy133) [inline modules/plink2/extract/main.nf:32]:
```suggestion
```

File: modules/plink2/extract/main.nf [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    if( "$pgen" == "${prefix}.pgen" ) error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"
+    """
+    plink2 \\
+        --pfile ${pgen.baseName} \\
+        $args \\
+        --threads $task.cpus \\
+        --extract $variants \\
+        --make-pgen \\
+        --out ${prefix}
```
Comment (jfy133) [inline modules/plink2/extract/main.nf:30]:
```suggestion
        --out ${prefix}
```

I'm not familiar with the tool, can downstream tools accept gzipped output from `plink2`? If not don't worry, if yes it woudl be great to have a gzip command after this one to compress all the output for publishing.

File: modules/plink2/extract/meta.yml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+  - extract
+tools:
+  - plink2:
+      description: |
+          Whole genome association analysis toolset, designed to perform a range
+          of basic, large-scale analyses in a computationally efficient manner
+      homepage: http://www.cog-genomics.org/plink/2.0/
+      documentation: http://www.cog-genomics.org/plink/2.0/general_usage
+      tool_dev_url: None
+      doi: ""
```
Comment (jfy133) [inline modules/plink2/extract/meta.yml:14]: Please add DOI from: `Chang CC, Chow CC, Tellier LCAM, Vattikuti S, Purcell SM, Lee JJ (2015)
Second-generation PLINK: rising to the challenge of larger and richer datasets.
GigaScience, 4.
`

(this paper is referenced in `plink2 --help`)

File: modules/plink2/extract/meta.yml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - versions:
+      type: file
+      description: File containing software versions
+      pattern: "versions.yml"
+  - pgen:
+      type: file
+      description: PLINK 2 binary genotype table
```
Comment (jfy133) [inline modules/plink2/extract/meta.yml:52]: For each of these, extend the sentence so to distinguish from the input

**Review by jfy133 — CHANGES_REQUESTED** [review]
A couple of minor things but overall looks pretty good to me (with the caveat I'm not familiar with the tool)

File: modules/plink2/extract/main.nf (hunk 2 above)
Comment (nebfield) [reply modules/plink2/extract/main.nf:32]: oops :see_no_evil:

File: modules/plink2/extract/main.nf (hunk 3 above)
Comment (nebfield) [reply modules/plink2/extract/main.nf:30]: `plink2` natively supports [zstandard compression](http://www.cog-genomics.org/plink/2.0/other#zstd) :thinking:

File: modules/plink2/extract/meta.yml (hunk 4 above)
Comment (nebfield) [reply modules/plink2/extract/meta.yml]: 7e782c0  :smile:

File: modules/plink2/extract/meta.yml (hunk 5 above)
Comment (nebfield) [reply modules/plink2/extract/meta.yml]: good point! are the new descriptions clearer? 40ca9ce

File: modules/plink2/extract/main.nf [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        'quay.io/biocontainers/plink2:2.00a2.3--h712d239_1' }"
+
+    input:
+    tuple val(meta), path(pgen), path(psam), path(pvar), path(variants)
+
+    output:
+    tuple val(meta), path("*.pgen")    , emit: extract_pgen
+    tuple val(meta), path("*.psam")    , emit: extract_psam
+    tuple val(meta), path("*.pvar.zst"), emit: extract_pvar
+    path "versions.yml"            , emit: versions
```
Comment (jfy133) [inline modules/plink2/extract/main.nf:17]:
```suggestion
    tuple val(meta), path("*.pvar.zst"), emit: extract_pvar
    path "versions.yml"                , emit: versions
```

**Review by jfy133 — APPROVED** [review]

**Pull request merged**